### PR TITLE
Simplify ETH Transaction

### PR DESCRIPTION
**Pull request by marktoda**
Remove the storage of "serialized" and "parsed" transactions separately,
by always storing a custom wrapper around an Ethereum tx object. This
allows us to simply serialize and deserialize the ethereum transaction
into whatever format, while avoiding the need to handle multiple cases
depending on which type of transaction we are currently holding

Ticket: BG-17980

### CODE REVIEW

File: src/coin/cgld/utils.ts [hunk 1]
```diff
@@ -12,7 +13,7 @@ import { SigningError } from '../baseCoin/errors';
  * @param {KeyPair} keyPair the signer's keypair
  * @returns {string} the transaction signed and encoded
  */
-export async function sign(transactionData: TxData, keyPair: KeyPair): Promise<string> {
+export async function sign(transactionData: TxJson, keyPair: KeyPair): Promise<string> {
```
Comment (argjv) [inline src/coin/cgld/utils.ts:16]: Update JSDocs

File: src/coin/eth/iface.ts [hunk 2]
```diff
@@ -1,23 +1,23 @@
...
-}
-
 export interface FieldStruct {
   components?: any;
   name: string;
   inputs?: any;
   type: string;
 }
+
+export interface TxJson {
```
Comment (argjv) [inline src/coin/eth/iface.ts:14]: Is this the same format transactions are seen in the network? If so, `ParsedTransaction` may be a better name.

Comment (marktoda) [reply src/coin/eth/iface.ts:14]: No, transactions on the network are seen in the fully serialized hex format. This is the format that the rest of BitGo usually plays well with (decimal string numbers, etc.)

File: src/coin/cgld/utils.ts (hunk 1 above)
Comment (marktoda) [reply src/coin/cgld/utils.ts:16]: done

File: src/coin/eth/types.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
+import BigNumber from 'bignumber.js';
+import { Transaction as EthereumTx } from 'ethereumjs-tx';
+import { addHexPrefix, bufferToHex, bufferToInt } from 'ethereumjs-util';
+import { TxJson } from './iface';
+
+/**
+ * an Ethereum transaction with helpers for serialization and deserialization.
+ */
+export class EthTx {
```
Comment (argjv) [inline src/coin/eth/types.ts:9]: s/`EthTx `/`EthTransaction`/g

File: src/coin/eth/types.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
+import BigNumber from 'bignumber.js';
+import { Transaction as EthereumTx } from 'ethereumjs-tx';
+import { addHexPrefix, bufferToHex, bufferToInt } from 'ethereumjs-util';
+import { TxJson } from './iface';
+
+/**
+ * an Ethereum transaction with helpers for serialization and deserialization.
```
Comment (argjv) [inline src/coin/eth/types.ts:7]: nit: capitalize first word

File: src/coin/eth/transaction.ts [hunk 5]
```diff
@@ -4,37 +4,45 @@
...
 import { KeyPair } from './keyPair';
-import { TxData } from './iface';
-import { Utils } from './';
+import { TxJson } from './iface';
+import { EthTx } from './types';
 
 export class Transaction extends BaseTransaction {
-  protected _parsedTransaction?: TxData; // transaction in JSON format
-  protected _encodedTransaction?: string; // transaction in hex format
+  protected _parsedTransaction?: EthTx;
```
Comment (argjv) [inline src/coin/eth/transaction.ts:13]: Rename `_parsedTransaction`, it is not a JSON anymore, but an object. `_ethTransaction`?

File: src/coin/cgld/iface.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+export interface CeloTxData {
```
Comment (argjv) [inline src/coin/cgld/iface.ts:1]: `TxData` is ok. You can always rename it in the import statement when you import multiple `TxData`

File: src/coin/cgld/utils.ts [hunk 7]
```diff
@@ -1,18 +1,19 @@
 import { signTransaction } from '@celo/contractkit/lib/utils/signing-utils';
 import { addHexPrefix } from 'ethereumjs-util';
 import { BigNumber } from 'bignumber.js';
-import { TxData } from '../eth/iface';
+import { TxJson } from '../eth/iface';
 import { KeyPair } from '../eth/keyPair';
 import { SigningError } from '../baseCoin/errors';
+import { CeloTxData } from './iface';
```
Comment (argjv) [inline src/coin/cgld/utils.ts:7]:
```
import { TxData as EthTxData } from '../eth/iface';
...
import { TxData as CeloTxData } from './iface';
```

**Comment by marktoda** [comment]
Thanks for the review Jorge. My hope is that this makes it simpler, both for the simtlix team now and for maintainers in the future here. I noticed a lot of the work / pain for these guys was ser/deser so this should get rid of that.

File: src/coin/eth/transaction.ts [hunk 8]
```diff
@@ -59,26 +67,29 @@ export class Transaction extends BaseTransaction {
...
+  toBroadcastFormat(): string {
+    if (!this._ethTransaction) {
+      throw new InvalidTransactionError('No transaction data to format');
     }
-    return this._encodedTransaction;
+    return this._ethTransaction.toSerialized();
   }
 
   /** @inheritdoc */
   toJson(): any {
```
Comment (argjv) [inline src/coin/eth/transaction.ts:89]: return type `TxJson`

**Pull request merged**